Enable more rules in ruff

We recently removed to ruff (https://github.com/mlflow/mlflow/pull/8873) to lint python files faster. We can enable more rules. This is a ticker for tracking that. To make PRs reviewable, we should enable rules one by one.

Rules that I think are useful:

- [x] https://beta.ruff.rs/docs/rules/redundant-open-modes/ (@EdAbati)
- [ ] https://beta.ruff.rs/docs/rules/#flake8-print-t20 (@Bncer)

Example PR:

https://github.com/mlflow/mlflow/pull/8958

Dear @harupy!
Thank you for the opportunity!
I have discovered unexpected behavior of [flake8-print-t20](https://beta.ruff.rs/docs/rules/#flake8-print-t20) 
For example in add_num.py file:
```
import os
def add_numbers(a, b):
    print(f"The sum of {a} and {b} is {a + b}")
    return a + b
```
 on command  - `ruff check add_num.py`
I have following ruff status:
```
add_num.py:1:8: F401 [*] os imported but unused
add_num.py:4:5: T201 `print` found
Found 2 errors.
[*] 1 potentially fixable with the --fix option.
```
_only 1 potentially fixable_
and on command - `ruff --fix .`
```
add_num.py:3:5: T201 `print` found
Found 2 errors (1 fixed, 1 remaining).
```
So ruff auto fix F401, but leave our T201 rule.
**It seems doesn't have autofix of T201**

As for ruff checking the whole project mlflow:
we have 25 errors mainly in `dev` folder, where files used in `./.github/workflows/`
As a proposition we can include `# noqa: T201` for each line to ignore prints or add `dev` folder in exclude list of pyproject.toml

--- a/pylintrc
+++ b/pylintrc
@@ -83,7 +83,6 @@ enable=signature-differs,
        # --------------
        # Custom rules
        pytest-raises-without-match,
-       print-function,
        unittest-assert-raises,
        pytest-raises-multiple-statements,
        use-f-string,

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -25,8 +25,7 @@ select = [
   "UP015",
   "UP031",
   "UP034",
-  "T201",
-  "T203",
+  "T20",
 ]
 force-exclude = true
 ignore = [

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -25,6 +25,8 @@ select = [
   "UP015",
   "UP031",
   "UP034",
+  "T201",
+  "T203",
 ]
 force-exclude = true
 ignore = [
@@ -70,3 +72,7 @@ exclude = [
   "mlflow/temporary_db_migrations_for_pre_1_users",
   "tests/protos",
 ]
+[tool.ruff.per-file-ignores]
+"dev/*" = ["T201", "T203"]
+"tests/resources/example_project/greeter.py" = ["T201", "T203"]
+"tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py" = ["T201", "T203"]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -73,6 +73,4 @@ exclude = [
   "tests/protos",
 ]
 [tool.ruff.per-file-ignores]
-"dev/*" = ["T201", "T203"]
-"tests/resources/example_project/greeter.py" = ["T201", "T203"]
-"tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py" = ["T201", "T203"]
+"dev/*" = ["T201"]

--- a/tests/resources/example_project/greeter.py
+++ b/tests/resources/example_project/greeter.py
@@ -14,4 +14,4 @@ if __name__ == "__main__":
     if args.excitement is not None:
         greeting.append("!" * args.excitement)
     # pylint: disable-next=print-function
-    print(" ".join(greeting))
+    print(" ".join(greeting))  # noqa: T201

--- a/tests/resources/example_project/greeter.py
+++ b/tests/resources/example_project/greeter.py
@@ -13,5 +13,4 @@ if __name__ == "__main__":
     greeting = [args.greeting, args.name]
     if args.excitement is not None:
         greeting.append("!" * args.excitement)
-    # pylint: disable-next=print-function
     print(" ".join(greeting))  # noqa: T201

--- a/tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py
+++ b/tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py
@@ -53,7 +53,7 @@ class PluginDeploymentClient(BaseDeploymentClient):
 
 def run_local(name, model_uri, flavor=None, config=None):
     # pylint: disable-next=print-function
-    print(
+    print(  # noqa: T201
         f"Deployed locally at the key {name} using the model from {model_uri}. "
         f"It's flavor is {flavor} and config is {config}"
     )

--- a/tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py
+++ b/tests/resources/mlflow-test-plugin/mlflow_test_plugin/fake_deployment_plugin.py
@@ -52,7 +52,6 @@ class PluginDeploymentClient(BaseDeploymentClient):
 
 
 def run_local(name, model_uri, flavor=None, config=None):
-    # pylint: disable-next=print-function
     print(  # noqa: T201
         f"Deployed locally at the key {name} using the model from {model_uri}. "
         f"It's flavor is {flavor} and config is {config}"

